programs/mango-v4/src/instructions/token_force_close_borrows_with_token.rs

```diff
@@ -0,0 +1,207 @@
+use anchor_lang::prelude::*;
+use fixed::types::I80F48;
+use std::cmp::min;
+
+use crate::accounts_ix::*;
+use crate::error::*;
+use crate::health::*;
+use crate::logs::TokenBalanceLog;
+use crate::state::*;
+
+// TODO: alternative ix name suggestions?
+pub fn token_force_close_borrows_with_token(
+    ctx: Context<TokenForceCloseBorrowsWithToken>,
+    asset_token_index: TokenIndex,
+    // token's mode is checked at #2
+    liab_token_index: TokenIndex,
+    max_liab_transfer: I80F48,
+) -> Result<()> {
+    let group_pk = &ctx.accounts.group.key();
+
+    require_neq!(asset_token_index, liab_token_index, MangoError::SomeError);
+    // TODO: should we enforce that asset token index is always USDC?
+
+    let mut account_retriever = ScanningAccountRetriever::new(ctx.remaining_accounts, group_pk)
+        .context("create account retriever")?;
+
+    require_keys_neq!(ctx.accounts.liqor.key(), ctx.accounts.liqee.key());
+    let mut liqor = ctx.accounts.liqor.load_full_mut()?;
+    // account constraint #1
+    require!(
+        liqor
+            .fixed
+            .is_owner_or_delegate(ctx.accounts.liqor_owner.key()),
+        MangoError::SomeError
+    );
+    require_msg_typed!(
+        !liqor.fixed.being_liquidated(),
+        MangoError::BeingLiquidated,
+        "liqor account"
+    );
+
+    let mut liqee = ctx.accounts.liqee.load_full_mut()?;
+    let mut liqee_health_cache = new_health_cache(&liqee.borrow(), &account_retriever)
+        .context("create liqee health cache")?;
+    let liqee_liq_end_health = liqee_health_cache.health(HealthType::LiquidationEnd);
```

Comment:
unclear why we compute this number here; if the whole liqee_health_cache is created further down, we don't need to bother adjusting it